### PR TITLE
Added functions for working with the Custom week number

### PR DESCRIPTION
**Pull request by andyyzh**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature
Short description (up to few sentences):
    The week number 1 is the first week in year that contains some day, which default to January 1, it can be modify by parameter of function. Week begins at monday.
    In this way, users can customize the first week of the year through parameters.

Detailed description (optional):
### toCustomWeek
Converts a date or date with time to a UInt8 number containing the Custom Week number.
toCustomWeek(DateTime, [, CustomDate][, Timezone])

### toCustomYear

Converts a date or date with time to a UInt16 number containing the Custom Year number.
toCustomYear(DateTime, [, CustomDate][, Timezone])

### toStartOfCustomYear
Rounds down a date or date with time to the first day of Custom year. Returns the date.
toStartOfCustomYear(DateTime, [, CustomDate][, Timezone])

### Example:
By default, the week contains January 1 is the first week:
SELECT toDate('2016-12-27') AS date, toCustomYear(date) AS year, toCustomWeek(date) AS week, toStartOfCustomYear(date) AS startday;

┌───────date─┬─year─┬─week─┬───startday─┐
│ 2016-12-27         │ 2017   │          1  │ 2016-12-26   │
└───────────┴─────┴─────┴─────────┘

The week contains January 28 is the first week:
SELECT toDate('2016-12-27') AS date, toCustomYear(date, '01-28') AS year, toCustomWeek(date,'01-28') AS week, toStartOfCustomYear(date,'01-28') AS startday;

┌───────date─┬─year─┬─week─┬───startday─┐
│ 2016-12-27         │ 2016   │        49  │ 2016-01-25  │
└───────────┴────┴─────┴─────────┘

### CODE REVIEW

**Comment by blinkov** [comment]
@andyyzh what's the usecase for this? is it somehow related to Lunar calendar?

**Comment by andyyzh** [comment]
> @andyyzh what's the usecase for this? is it somehow related to Lunar calendar?

A week number computing rule: the week contained January 1st is the first week, contained a day to make configurable.
It has nothing to do with the lunar calendar.

> USA, Canada, most of Latin America, Japan, Israel, South Korea, among others, use a week numbering system (called North American in our Calculator) in which the first week (numbered 1) of any given year is the week which contains January 1st. 
[https://www.calendar-12.com/week_number](url)

**Comment by alexey-milovidov** [comment]
I'm trying to figure out the motivation for this implementation. (Sorry, it doesn't look clear for me.)

It looks like that the initial motivation is to provide week numbering to start first week from January 1st.

But you probably have noticed that the ISO week numbering starts from the first week containing January 4th. And you wanted to generalize this function, so you've added the possibility to start the numbering from any selected day within a year. Is this really needed?

I'm asking because it looks like a false generalization. ISO week numbering starts from the first week containing January 4th, because, coincidentally, it is also the first week that contains majority of days (at least 4 of 7) within a year.

And the possibility to start from the first week containing January 3rd or January 5th doesn't make much sense. It is still can make sense to start from a week containing January 1st (first week with at least one day within year) or January 7th (first full week within a year).

But I wonder if it is motivated by some unusual local calendars...

**Comment by alexey-milovidov** [comment]
For week numbering system, at least the following customizations make sense:
- start a week from Monday (as in Europe) or Sunday (as in USA) or Saturday;
- allow numbered week to cross the year boundary (as in ISO-8601); or reset the numbering at January 1st; or use number zero instead of one if a week belongs to previous year;
- how to select the first week of a year: by first day of year or by majority of days within a year (ISO-8601) or containing completely within a year. 

For example of (still incomplete) generalization of week numbering system, look at:
https://dev.mysql.com/doc/refman/5.5/en/date-and-time-functions.html#function_week
and 
https://en.wikipedia.org/wiki/Week#Week_numbering

**Comment by andyyzh** [comment]
> But you probably have noticed that the ISO week numbering starts from the first week containing January 4th. And you wanted to generalize this function, so you've added the possibility to start the numbering from any selected day within a year. Is this really needed?
toCustomWeek(DateTime, [, CustomDate][, Timezone])
Adding 'CustomDate' parameter is unnecessary, it's to make functions more general.

> But I wonder if it is motivated by some unusual local calendars...
No special calendar requirement,  

I plan to delete the 'CustomDate' parameter to make the purpose of the function clearer.

Do you want to change the name of the function?  called NorthWeek or USWeek?

**Comment by alexey-milovidov** [comment]
> I plan to delete the 'CustomDate' parameter to make the purpose of the function clearer.
Do you want to change the name of the function? called NorthWeek or USWeek?

Ok. Yes, better to change the name.
USWeek is not appropriate, because weeks start at Sunday in US.
NorthWeek is too confusing.

Better to implement the `week` function from MySQL (at least for compatibility).
https://dev.mysql.com/doc/refman/5.5/en/date-and-time-functions.html#function_week

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

**Comment by andyyzh** [comment]
Prepare to add these functions that similar to mysql, and add new mode:

> WEEK(date[,mode])
YEAR(date[,mode])
YEARWEEK(date[,mode])

I'll commit it in two weeks.

**Comment by alexey-milovidov** [comment]
Ok.

**Comment by andyyzh** [comment]
Implementing functions similar to those in MySQL;
## week(date[,mode])
This function returns the week number for date or datetime. The two-argument form of week() enables you to specify whether the week starts on Sunday or Monday and whether the return value should be in the range from 0 to 53 or from 1 to 53. If the mode argument is omitted, the default mode is 0.
The following table describes how the mode argument works.

| Mode | First day of week | Range |  Week 1 is the first week … |
| ----------- | -------- | -------- | ------------------ |
|0|Sunday|0-53|with a Sunday in this year
|1|Monday|0-53|with 4 or more days this year
|2|Sunday|1-53|with a Sunday in this year
|3|Monday|1-53|with 4 or more days this year
|4|Sunday|0-53|with 4 or more days this year
|5|Monday|0-53|with a Monday in this year
|6|Sunday|1-53|with 4 or more days this year
|7|Monday|1-53|with a Monday in this year
|8|Sunday|1-53|contains January 1
|9|Monday|1-53|contains January 1

For mode values with a meaning of “with 4 or more days this year,” weeks are numbered according to ISO 8601:1988:

- If the week containing January 1 has 4 or more days in the new year, it is week 1.

- Otherwise, it is the last week of the previous year, and the next week is week 1.

For mode values with a meaning of “contains January 1”, the week contains January 1 is week 1. It doesn't matter how many days in the new year the week contained, even if it contained only one day.

```
week(DateTime1, [, mode][, Timezone])

```
**Parameters**

- `DateTime1` – Date or DateTime.
- `mode` – Optional parameter, Range of values is [0,9], default is 0.
- `Timezone` –  Optional parameter, it behaves like any other conversion function.

**Example**

``` sql
SELECT toDate('2016-12-27') AS date, week(date) AS week0, week(date,1) AS week1, week(date,9) AS week9;
```

```
┌───────date─┬─week0─┬─week1─┬─week9─┐
│ 2016-12-27 │    52 │    52 │     1 │
└────────────┴───────┴───────┴───────┘
```

## yearWeek(date[,mode])
Returns year and week for a date. The year in the result may be different from the year in the date argument for the first and the last week of the year.

The mode argument works exactly like the mode argument to week(). For the single-argument syntax, a mode value of 0 is used. 

**Example**

``` sql
SELECT toDate('2016-12-27') AS date, yearWeek(date) AS yearWeek0, yearWeek(date,1) AS yearWeek1, yearWeek(date,9) AS yearWeek9;
```

```
┌───────date─┬─yearWeek0─┬─yearWeek1─┬─yearWeek9─┐
│ 2016-12-27 │    201652 │    201652 │    201701 │
└────────────┴───────────┴───────────┴───────────┘
```

**Comment by andyyzh** [comment]
@alexey-milovidov  
Does default mode value need to be added to the system settings?
FYI: 
[https://dev.mysql.com/doc/refman/5.5/en/server-system-variables.html#sysvar_default_week_format](https://dev.mysql.com/doc/refman/5.5/en/server-system-variables.html#sysvar_default_week_format)

**Comment by andyyzh** [comment]
@vitlibar ,  I'm done.  please review code, thank you.

File: docs/en/query_language/functions/date_time_functions.md [hunk 1]
```diff
@@ -169,6 +169,77 @@ Converts a date or date with time to a UInt16 number containing the ISO Year num
 
 Converts a date or date with time to a UInt8 number containing the ISO Week number.
 
+## toCustomYear
+
+Converts a date or date with time to a UInt16 number containing the Custom Year number.
+```
+toCustomYear(DateTime, [, CustomDate][, Timezone])
```
Comment (vitlibar) [inline docs/en/query_language/functions/date_time_functions.md]: It seems your PR doesn't contain the function `toCustomYear` anymore.

File: docs/en/query_language/functions/date_time_functions.md [hunk 2]
```diff
@@ -169,6 +169,77 @@ Converts a date or date with time to a UInt16 number containing the ISO Year num
...
+For mode values with a meaning of “with 4 or more days this year,” weeks are numbered according to ISO 8601:1988:
+
+- If the week containing January 1 has 4 or more days in the new year, it is week 1.
+
+- Otherwise, it is the last week of the previous year, and the next week is week 1.
+
+For mode values with a meaning of “contains January 1”, the week contains January 1 is week 1. It doesn't matter how many days in the new year the week contained, even if it contained only one day.
+
+```
+week(DateTime1, [, mode][, Timezone])
```
Comment (vitlibar) [inline docs/en/query_language/functions/date_time_functions.md]: Too complicated name for the documentation. Please rename `DateTime1` => `date`

File: dbms/src/Functions/FunctionCustomWeekToSomething.h [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        }
+        else
+            throw Exception(
+                "Number of arguments for function " + getName() + " doesn't match: passed " + toString(arguments.size())
+                    + ", should be 1 or 2 or 3",
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        /// For DateTime, if time zone is specified, attach it to type.
+        if (std::is_same_v<ToDataType, DataTypeDateTime>)
+            return std::make_shared<DataTypeDateTime>(extractTimeZoneNameFromFunctionArguments(arguments, 1, 0));
```
Comment (vitlibar) [inline dbms/src/Functions/FunctionCustomWeekToSomething.h]: It's not very clear. The class template `FunctionCustomWeekToSomething` is used only to implement the `week` and `yearWeek` functions, isn't it? `week()` returns an integer 0..53, `yearWeek` returns a string `yyyyww`. So it seems the return type should be `DataTypeUInt8` or `DataTypeUInt16`, not `DataTypeDateTime`.

File: docs/en/query_language/functions/date_time_functions.md (hunk 1 above)
Comment (andyyzh) [reply docs/en/query_language/functions/date_time_functions.md]: Yes, it needs to be deleted.

File: dbms/src/Functions/FunctionCustomWeekToSomething.h (hunk 3 above)
Comment (andyyzh) [reply dbms/src/Functions/FunctionCustomWeekToSomething.h]: I'll fix it right away, 
`week `return DataTypeUInt8    `yearWeek `return DataTypeUInt32

File: docs/en/query_language/functions/date_time_functions.md [hunk 4]
```diff
@@ -169,6 +169,77 @@ Converts a date or date with time to a UInt16 number containing the ISO Year num
...
 Converts a date or date with time to a UInt8 number containing the ISO Week number.
 
+## toCustomYear
+
+Converts a date or date with time to a UInt16 number containing the Custom Year number.
+```
+toCustomYear(DateTime, [, CustomDate][, Timezone])
+```
+
+## week(date[,mode])
```
[the remark this answers is not in the document]
Comment (andyyzh) [reply docs/en/query_language/functions/date_time_functions.md]: ok

File: docs/en/query_language/functions/date_time_functions.md (hunk 2 above)
Comment (andyyzh) [reply docs/en/query_language/functions/date_time_functions.md]: ok

**Comment by andyyzh** [comment]
All checks have passed.

**Comment by andyyzh** [comment]
All checks have passed.

File: docs/en/query_language/functions/date_time_functions.md [hunk 5]
```diff
@@ -169,6 +169,72 @@ Converts a date or date with time to a UInt16 number containing the ISO Year num
...
+|0|Sunday|0-53|with a Sunday in this year
+|1|Monday|0-53|with 4 or more days this year
+|2|Sunday|1-53|with a Sunday in this year
+|3|Monday|1-53|with 4 or more days this year
+|4|Sunday|0-53|with 4 or more days this year
+|5|Monday|0-53|with a Monday in this year
+|6|Sunday|1-53|with 4 or more days this year
+|7|Monday|1-53|with a Monday in this year
+|8|Sunday|1-53|contains January 1
+|9|Monday|1-53|contains January 1
```
Comment (vitlibar) [inline docs/en/query_language/functions/date_time_functions.md]: I didn't see the modes `8` and `9` in the [MySQL documentation](https://dev.mysql.com/doc/refman/5.5/en/date-and-time-functions.html#function_week). Did you take them from somewhere else?

File: libs/libcommon/include/common/DateLUTImpl.h [hunk 6]
```diff
@@ -14,10 +14,17 @@
...
+#define WEEK_NEWYEAR_DAY 8
+
+
 #if defined(__PPC__)
-#if !__clang__
-#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
-#endif
+#    if !__clang__
+#        pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#    endif
```
Comment (vitlibar) [inline libs/libcommon/include/common/DateLUTImpl.h]: Why did you change this?

File: libs/libcommon/include/common/DateLUTImpl.h [hunk 7]
```diff
@@ -14,10 +14,17 @@
 #define DATE_LUT_MAX_YEAR 2105 /// Last supported year
 #define DATE_LUT_YEARS (1 + DATE_LUT_MAX_YEAR - DATE_LUT_MIN_YEAR) /// Number of years in lookup table
 
+/// Flags for calc_week() function.
+#define WEEK_MONDAY_FIRST 1
+#define WEEK_YEAR 2
+#define WEEK_FIRST_WEEKDAY 4
+#define WEEK_NEWYEAR_DAY 8
```
Comment (vitlibar) [inline libs/libcommon/include/common/DateLUTImpl.h]: I suppose it would be much better to use `enum class WeakMode` (with constants from 0 to 9) here instead of these old-styled macros.

File: libs/libcommon/include/common/DateLUTImpl.h [hunk 8]
```diff
@@ -352,54 +313,205 @@ class DateLUTImpl
...
+            {
+                (*year)++;
+                return 1;
+            }
+        }
+        return days / 7 + 1;
     }
 
-    inline unsigned toISOWeek(time_t t) const
+    inline unsigned calc_yearWeek(DayNum d, UInt32 week_mode) const
```
Comment (vitlibar) [inline libs/libcommon/include/common/DateLUTImpl.h]: It looks like you don't need two functions `calc_week()` and `calc_yearWeek()` because you have to calculate `year` anyway. So it seems it's enough to have only function

```
std::pair<UInt16 /* year */, UInt8 /* week */> toYearAndWeek(DayNum d, WeekMode week_mode) const
```

and the `WeekImpl::execute()` and `YearWeekImpl::execute()` functions get from this pair everything they need.

File: libs/libcommon/include/common/DateLUTImpl.h [hunk 9]
```diff
@@ -352,54 +313,205 @@ class DateLUTImpl
...
-        return toRelativeMonthNum(toDayNum(t));
+        UInt32 week_format = (mode & 7);
+        if (!(week_format & WEEK_MONDAY_FIRST))
+            week_format ^= WEEK_FIRST_WEEKDAY;
+        return week_format;
     }
 
-    inline unsigned toRelativeQuarterNum(DayNum d) const
+    /*
+      Calculate nr of day since year 0 in new date-system (from 1615)
```
Comment (vitlibar) [inline libs/libcommon/include/common/DateLUTImpl.h]: What's the purpose of this fuinction? We already have `lut` and `years_lut` arrays which allow to do datetime calculations very quickly. And we don't need to support years from 1615 while we are storing dates as an `uint16` number of days since Unix epoch.

File: libs/libcommon/include/common/DateLUTImpl.h [hunk 10]
```diff
@@ -118,15 +121,9 @@ class DateLUTImpl
         return lut[DayNum(index - (lut[index].day_of_week - 1))].date;
     }
 
-    inline DayNum toFirstDayNumOfWeek(DayNum d) const
-    {
-        return DayNum(d - (lut[d].day_of_week - 1));
-    }
+    inline DayNum toFirstDayNumOfWeek(DayNum d) const { return DayNum(d - (lut[d].day_of_week - 1)); }
```
Comment (vitlibar) [inline libs/libcommon/include/common/DateLUTImpl.h]: For consistency the functions `toFirstDayNumOfWeek()` and `toStartOfWeekInterval()` should get the new parameter `week_mode` too to find out whether weeks starts on Sunday or on Monday.

File: docs/en/query_language/functions/date_time_functions.md (hunk 5 above)
Comment (andyyzh) [reply docs/en/query_language/functions/date_time_functions.md]: FYI:
https://stackoverflow.com/questions/55658435/python-north-american-work-week-number-from-datetime
https://www.calendar-12.com/week_number

File: libs/libcommon/include/common/DateLUTImpl.h (hunk 6 above)
Comment (andyyzh) [reply libs/libcommon/include/common/DateLUTImpl.h]: I didn't change it manually, automatically modified by the clang-format.

File: dbms/tests/queries/0_stateless/00941_to_custom_week.sql [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+SELECT yearWeek(toDate('2000-01-06'),0) AS w2000, yearWeek(toDate('2001-01-06'),0) AS w2001, yearWeek(toDate('2002-01-06'),0) AS w2002,yearWeek(toDate('2003-01-06'),0) AS w2003, yearWeek(toDate('2004-01-06'),0) AS w2004, yearWeek(toDate('2005-01-06'),0) AS w2005, yearWeek(toDate('2006-01-06'),0) AS w2006;
+SELECT yearWeek(toDate('2000-01-01'),1) AS w2000, yearWeek(toDate('2001-01-01'),1) AS w2001, yearWeek(toDate('2002-01-01'),1) AS w2002,yearWeek(toDate('2003-01-01'),1) AS w2003, yearWeek(toDate('2004-01-01'),1) AS w2004, yearWeek(toDate('2005-01-01'),1) AS w2005, yearWeek(toDate('2006-01-01'),1) AS w2006;
+SELECT yearWeek(toDate('2000-01-06'),1) AS w2000, yearWeek(toDate('2001-01-06'),1) AS w2001, yearWeek(toDate('2002-01-06'),1) AS w2002,yearWeek(toDate('2003-01-06'),1) AS w2003, yearWeek(toDate('2004-01-06'),1) AS w2004, yearWeek(toDate('2005-01-06'),1) AS w2005, yearWeek(toDate('2006-01-06'),1) AS w2006;	
+SELECT week(toDate('1998-12-31'),2),week(toDate('1998-12-31'),3), week(toDate('2000-01-01'),2), week(toDate('2000-01-01'),3);
+SELECT week(toDate('2000-12-31'),2),week(toDate('2000-12-31'),3);
```
Comment (vitlibar) [inline dbms/tests/queries/0_stateless/00941_to_custom_week.sql]: Almost all the date/time functions in ClickHouse [have prefix "to"](https://clickhouse.yandex/docs/en/query_language/functions/date_time_functions/). For consistency it would be nice to rename `week()` => `toWeek()`, and `yearWeek()` => `toYearWeek()`, and add case-insensitive aliases `week`, `yearWeek` (like [this](https://github.com/yandex/ClickHouse/blob/febca43a894ff0a7cacfc51a16ad4a87bf5b10d2/dbms/src/Functions/FunctionsRound.cpp#L16)) for compatibility with MySQL.

File: docs/en/query_language/functions/date_time_functions.md (hunk 5 above)
Comment (vitlibar) [reply docs/en/query_language/functions/date_time_functions.md]: thanks

File: dbms/tests/queries/0_stateless/00941_to_custom_week.sql (hunk 11 above)
Comment (andyyzh) [reply dbms/tests/queries/0_stateless/00941_to_custom_week.sql]: I was thinking of supporting mysql syntax.
I'm going to change it to week() => toWeek(), and yearWeek() => toYearWeek().

File: libs/libcommon/include/common/DateLUTImpl.h (hunk 7 above)
Comment (andyyzh) [reply libs/libcommon/include/common/DateLUTImpl.h]: ok, I'll fix it.

File: libs/libcommon/include/common/DateLUTImpl.h (hunk 8 above)
Comment (andyyzh) [reply libs/libcommon/include/common/DateLUTImpl.h]: ok, I'll fix it.

File: libs/libcommon/include/common/DateLUTImpl.h (hunk 6 above)
Comment (vitlibar) [reply libs/libcommon/include/common/DateLUTImpl.h]: clang-format can format only your changes

File: libs/libcommon/include/common/DateLUTImpl.h (hunk 9 above)
Comment (andyyzh) [reply libs/libcommon/include/common/DateLUTImpl.h]: This code is ported from mysql and I'm trying to change it.

File: libs/libcommon/include/common/DateLUTImpl.h (hunk 10 above)
Comment (andyyzh) [reply libs/libcommon/include/common/DateLUTImpl.h]: Let me do it.

Comment (andyyzh) [reply libs/libcommon/include/common/DateLUTImpl.h]: @vitlibar 
The default value of week_mode is 0, which is the first day of the week on Sunday.
`toFirstDayNumOfWeek()` is called `toMonday` for users, and if you add new parameter week_mode, it will output Sunday or Monday, which is confusing.

[https://clickhouse.yandex/docs/en/query_language/functions/date_time_functions/#tomonday](https://clickhouse.yandex/docs/en/query_language/functions/date_time_functions/#tomonday)
``` c++
struct ToMondayImpl
{
    static constexpr auto name = "toMonday";

    static inline UInt16 execute(UInt32 t, const DateLUTImpl & time_zone)
    {
        return time_zone.toFirstDayNumOfWeek(time_zone.toDayNum(t));
    }
```

Does default week_mode value need to be added to the system settings? 
The default value is recommended to be 3, so that we can remain the consistency of the original implementation.

toStartOfWeekInterval() itself has several parameters, isn't it too complicated for users to add week_mode?  There is no week interval calculation in the document.
[https://clickhouse.yandex/docs/en/query_language/functions/date_time_functions/#tostartofinterval-time-or-data-interval-x-unit-time-zone](https://clickhouse.yandex/docs/en/query_language/functions/date_time_functions/#tostartofinterval-time-or-data-interval-x-unit-time-zone)
`toStartOfInterval(time_or_data, INTERVAL x unit [, time_zone])`

**Summary:**
1. add new function：`toStartOfWeek(t[,mode])` to find out whether week start on Sunday or on Monday.
2. Add a `week_mode `configuration item to the system settings, default value of week_mode is 3.
3. `toStartOfInterval `does not change any more. 

Can you see if this is ok?

**Comment by andyyzh** [comment]
All previous tests have passed. I don't think the failing tests are caused by my change.
@alexey-milovidov   If you have time, take a look.

**Comment by andyyzh** [comment]
@vitlibar 
All the review issues have been fixed except toStartOfWeekInterval() function was not added.
Please review again.

**Comment by alexey-milovidov** [comment]
@vitlibar went on vacation. I will reassign this task to someone else.

**Comment by andyyzh** [comment]
> @vitlibar went on vacation. I will reassign this task to someone else.

ok, thanks!

File: libs/libcommon/include/common/DateLUTImpl.h (hunk 10 above)
Comment (vitlibar) [reply libs/libcommon/include/common/DateLUTImpl.h]: Ok, looks good. Why do you think the default value for `default_week_mode` should be `3`? In MySQL it's `0`.
